clique/src/lib.rs

```diff
@@ -181,3 +184,137 @@ impl Listener for TcpListener {
         Ok(stream)
     }
 }
+
+pub struct RateLimitedAsyncReadWrite<A> {
+    rate_limiter: SleepingRateLimiter,
+    read: A,
+}
+
+impl<A> RateLimitedAsyncReadWrite<A> {
+    pub fn new(read: A, rate_limiter: SleepingRateLimiter) -> Self {
+        Self { rate_limiter, read }
+    }
+}
+
+impl<A: AsyncRead + Unpin> AsyncRead for RateLimitedAsyncReadWrite<A> {
+    fn poll_read(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &mut tokio::io::ReadBuf<'_>,
+    ) -> std::task::Poll<std::io::Result<()>> {
+        let this = self.get_mut();
+        if Pin::new(&mut this.rate_limiter.current_sleep())
+            .poll(cx)
+            .is_pending()
+        {
+            return std::task::Poll::Pending;
+        }
+
+        let filled_before = buf.filled().len();
+        let result = Pin::new(&mut this.read).poll_read(cx, buf);
+        let filled_after = buf.filled().len();
+        let last_read_size = filled_after - filled_before;
+
+        this.rate_limiter.rate_limit(last_read_size);
+
+        result
+    }
+}
+
+impl<A: AsyncWrite + Unpin> AsyncWrite for RateLimitedAsyncReadWrite<A> {
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &[u8],
+    ) -> std::task::Poll<Result<usize, std::io::Error>> {
+        Pin::new(&mut self.read).poll_write(cx, buf)
+    }
+
+    fn poll_flush(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), std::io::Error>> {
+        Pin::new(&mut self.read).poll_flush(cx)
+    }
+
+    fn poll_shutdown(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), std::io::Error>> {
+        Pin::new(&mut self.read).poll_shutdown(cx)
+    }
+}
+
+impl<R: Splittable> Splittable for RateLimitedAsyncReadWrite<R> {
```

Comment:
nit: R -> A, as everywhere else you had used A. Or use better name like `Read|Reader` or smth like this

Reply:
removed